src/pb/publisher.rs

```diff
@@ -2,45 +2,131 @@
 // SPDX-FileCopyrightText: © Contributors to the iceoryx-rs project
 // SPDX-FileContributor: Mathias Kraus
 
-use super::{sample::SampleMut, Topic, POD};
-use crate::IceOryxError;
+use super::{ffi::Publisher as FfiPublisher, sample::SampleMut, PublisherOptions, POD};
+use crate::ConsumerTooSlowPolicy;
+use crate::IceoryxError;
+
+use std::marker::PhantomData;
+
+pub struct PublisherBuilder<'a, T: POD> {
+    service: &'a str,
+    instance: &'a str,
+    event: &'a str,
+    options: PublisherOptions,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T: POD> PublisherBuilder<'a, T> {
+    pub fn new(service: &'a str, instance: &'a str, event: &'a str) -> Self {
+        Self {
+            service,
+            instance,
+            event,
+            options: PublisherOptions::default(),
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn history_capacity(mut self, history_capacity: u64) -> Self {
+        self.options.history_capacity = history_capacity;
+        self
+    }
+
+    pub fn node_name(mut self, node_name: String) -> Self {
+        self.options.node_name = node_name;
+        self
+    }
+
+    pub fn subscriber_too_slow_policy(
+        mut self,
+        subscriber_too_slow_policy: ConsumerTooSlowPolicy,
+    ) -> Self {
+        self.options.subscriber_too_slow_policy = subscriber_too_slow_policy;
+        self
+    }
+
+    pub fn create(mut self) -> Result<Publisher<T>, IceoryxError> {
+        self.options.offer_on_create = true;
+        let ffi_pub = FfiPublisher::new(self.service, self.instance, self.event, &self.options)
+            .ok_or(IceoryxError::PublisherCreationFailed)?;
+
+        Ok(Publisher {
+            ffi_pub,
+            phantom: PhantomData,
+        })
+    }
+
+    pub fn create_without_offer(mut self) -> Result<InactivePublisher<T>, IceoryxError> {
```

Comment:
Could you please just have one unique `create` and add an option `offer_on_create` to the builder?

Reply:
No, since it is not possible to use the type state pattern ... or at least it would be cumbersome for the user since one would have to use enums, e.g.

```Rust
enum Publisher {
  Inactive(InactivePublisher),
  Active(ActivePublisher),
}

let publisher = PublisherBuilder::<u64>::new("All", "Glory", "Hypnotoad").create();

match publisher {
  Publisher::Active(p) => p.loan ...
}
```
It's getting nasty. Multiple create methods solve this nicely.

Reply:
Okay now I understand your problem. But the problem now is that you require a lot of create function variations and they are becoming exponentially more if we introduce more options in iceoryx which restrict the interface.

Reply:
Well, this is the idiomatic way. Maybe once [specialization](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md) is stabilized it could be used to simplify this.